c2wwise rewrite, fixes #123, #126, and #127

eu-ch opened the pull request:
Summary
- Almost complete c2wwise rewrite: new plugin API and build system
- Added support for output events
- Added support for 5.1, 7.1, and 7.1.4 (Dolby) channel configuration
- Updated docs

dromer commented on docs/03.gen.wwise.md:
```diff
@@ -24,176 +48,53 @@ Exposing an input parameter will automatically generate a Slider UI in the plugi
+Heavy generates a project compatible with Audiokinetic' [Development Tools](https://www.audiokinetic.com/en/library/edge/?source=SDK&id=effectplugin_tools.html). They are thoroughly documented, so we'll only provide one example.
 
-First create a new Wwise project, add a Source plugin to a game object, create an Event to trigger it and add it to a Soundbank.
+**Important:** Please close any running Wwise Authoring instance as it'll prevent pluging from being built.
```
minor typo -> `plugin` ;)

dromer commented on hvcc/generators/c2wwise/templates/additional_artifacts.json:
This is just a required file for the wwise build script?

If it's not a template maybe move it to a `static/` folder instead.

dromer commented:
Nice! this will require some testing of course. A basic example project could help with this.

Would you be able to create a super basic Wwise plugin that integrates with a really minimal Unity project perhaps?
So that we can try builds for different platforms and have something that actually be tested.

I'll do my best to try a build using this at least.

dromer commented:
Had to do some juggling with the SDK to get it to start building under Linux (spaces and special characters in path are evil; also executable flags for binaries are not set ..)

But I run into these errors:

```gcc
In file included from send_paramSource.cpp:33:
./send_paramSource.h:54:14: warning: 'Term' overrides a member function but is not marked 'override' [-Winconsistent-missing-override]
    AKRESULT Term(AK::IAkPluginMemAlloc* in_pAllocator);
             ^
../../../Linux_SDK_2022.1.8.8316/SDK/include/AK/SoundEngine/Common/IAkPlugin.h:701:20: note: overridden virtual function is here
                virtual AKRESULT Term(
                                 ^
send_paramSource.cpp:116:21: error: no member named 'm_EventQueue' in 'send_paramSource'
        in_pPlugin->m_EventQueue.Enqueue(EventID);
        ~~~~~~~~~~  ^
send_paramSource.cpp:191:60: error: use of undeclared identifier 'OnGlobalCallback'
    m_pWwiseCtx->GlobalContext()->UnregisterGlobalCallback(OnGlobalCallback, AkGlobalCallbackLocation_BeginRender);
                                                           ^
1 warning and 2 errors generated.
make[1]: *** [send_param_Linux_static.make:667: ../../../Linux_SDK_2022.1.8.8316/SDK/Linux_x64/Release/obj/send_paramSource/send_paramSource.o] Error 1
make: *** [send_param_Linux.make:88: send_paramSource] Error 2
```

eu-ch commented:
Thanks mate, I’ll take a look at the issues.Can you please send a PD patch that failed compilation? Also, can you elaborate on the pathing issues you’ve encountered on Linux?On Nov 5, 2023, at 7:44 AM, dreamer ***@***.***> wrote:﻿
Had to do some juggling with the SDK to get it to start building under Linux (spaces and special characters in path are evil; also executable flags for binaries are not set ..)
But I run into these errors:
In file included from send_paramSource.cpp:33:
./send_paramSource.h:54:14: warning: 'Term' overrides a member function but is not marked 'override' [-Winconsistent-missing-override]
    AKRESULT Term(AK::IAkPluginMemAlloc* in_pAllocator);
             ^
../../../Linux_SDK_2022.1.8.8316/SDK/include/AK/SoundEngine/Common/IAkPlugin.h:701:20: note: overridden virtual function is here
                virtual AKRESULT Term(
                                 ^
send_paramSource.cpp:116:21: error: no member named 'm_EventQueue' in 'send_paramSource'
        in_pPlugin->m_EventQueue.Enqueue(EventID);
        ~~~~~~~~~~  ^
send_paramSource.cpp:191:60: error: use of undeclared identifier 'OnGlobalCallback'
    m_pWwiseCtx->GlobalContext()->UnregisterGlobalCallback(OnGlobalCallback, AkGlobalCallbackLocation_BeginRender);
                                                           ^
1 warning and 2 errors generated.
make[1]: *** [send_param_Linux_static.make:667: ../../../Linux_SDK_2022.1.8.8316/SDK/Linux_x64/Release/obj/send_paramSource/send_paramSource.o] Error 1
make: *** [send_param_Linux.make:88: send_paramSource] Error 2


—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>

dromer commented:
When I install the path with wine it is `~/.wine/drive_c/Program Files (x86)/Audiokinetic/Wwise 2022.1.8.8316` this includes spaces ` ` and special characters `( )` which causes problems with the Wwise scripts.

So I moved the SDK somewhere else and named it `Wwise_Linux_SDK_2022.1.8.8316` instead.
Then at least `Tools/Linux/bin/premake5` needs to be made executable or you get `Permission denied` errors.

eu-ch commented:
Ah yes, I never use default installation paths myself, probably for this reason :)On Nov 5, 2023, at 9:24 AM, dreamer ***@***.***> wrote:﻿
When I install the path with wine it is ~/.wine/drive_c/Program Files (x86)/Audiokinetic/Wwise 2022.1.8.8316 this includes spaces   and special characters ( ) which causes problems with the Wwise scripts.
So I moved the SDK somewhere else and named it Wwise_Linux_SDK_2022.1.8.8316 instead.
Then at least Tools/Linux/bin/premake5 needs to be made executable or you get Permission denied errors.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>

eu-ch commented:
I fixed the PD compile patch. Will prepare a Unity project later during the week. Btw, do you plan to version bump on master branch some time soon? I can ensure I test with Unity before taht.

dromer commented:
What do you mean? When a new release is warranted a new release will be made.

Might do a tiny bug-release first. This is kind of a big change, so it would be good if some more people get a chance to test it.

dromer commented:
Btw with the commands I gave above I was able to successfully get a library file built!

However I'm not 100% on the UX of moving the file way from the build dir.
Now it is not clear there was any result from the build. If one wants to ship this in a project I assume one would have to place this file somewhere else again.

Besides: the authorization application doesn't run under Linux, so moving it to the `SDK/Linux_x86/Release/bin/` folder is kind of weird.

Or is this output folder automatically set by the Wwise scripts?
(I really find their whole stack confusing and unclear)

eu-ch commented:
That's great news!

Yeah, build artifacts are placed under SDK by convention.

> I really find their whole stack confusing and unclear
On the contrary, I find it quite reasonable :) Considering that my goal is to build a plugin and link it against a game across different platforms -- easy to build tools when everything is in predictable places. And for Heavy too, as it only needs to generate AK project files, and vendor tools will take over the build and packaging.

dromer commented:
Btw I'm expecting a PR from a company that did some customizations to c2wwise for their needs. Mainly relating to number of i/o. Might be interesting to look at extending the c2wwise capabilities to include something like that (for things like surround sound).

Relevant ticket:  https://github.com/Wasted-Audio/hvcc/issues/127

eu-ch commented:
I’m curious how they solved the multichannel problem in c2wwise. Wwise plugins can be placed anywhere in the pipeline and be presented with all kinds of channel configurations, and configurations can change in runtime, e.g. if you plug headphones into a PS5 controller. So, a pure data patch should adapt to it as well, which is impossible as all the patch graph gets statically compiled by Heavy.

So best case is that the patch can preallocate maximum number of I/O channels, and the Wwise plugin tries to ignore the unused ones. I bet this requires modifying API that HeavyContext uses, so that it doesn’t write into memory for, say, channels 3 and 4 when Wwise expects just two channels.

Or, you can build a 5.1 plugin which will refuse to be initialised in a different channel config. This can be alleviated by forcing a bus into 5.1, in which case Wwise will be up/downmixing busses before and after, depending on the situation. From the article you’ve linked in #127 this seems to be exactly what they’re doing:

> However, we found that the bus hosting the effect had to be limited to 5.1 in the Channel Configuration dropdown in Wwise to avoid serious stability issues, manifesting as crashes when using output devices with more than 6 channels.

Tbh, I’m surprised they didn’t have problems with channel configurations less than 6 channels, as I’d expect Heavy to write beyond allocated memory.

dromer commented:
That would depend on if your plugin is running at the end-stage or is just in the middle of the graph generating/processing a sound. No?

It's not like you need multi-channel wav files on order to play sampled sounds.

eu-ch commented:
Yeah, “Source” plugins can dictate channel configuration from Init method, I remember setting them to mono explicitly back when I was working on some internal samplers. I’m it so sure about “Effect” plugins though, and the following quote from the AK docs implies they can not:

>  A plug-in should handle several channel configurations (at least mono, stereo and 5.1 if it can be inserted on busses) or return AK_UnsupportedChannelConfig at inialization time.

Which kinda makes sense, as bus channel configuration is controlled by output device, which can change at runtime, and the whole pipeline must adapts. They could, in theory, recognise plugins that don’t fit and up/downmix appropriately, but I guess this edge case was considered unworthy of supporting.

So yeah, I think if we want to support more than two channels, we can impose a limitation that such plugins won’t work under any other channel configuration, and will produce an error and be bypassed in runtime. The workaround can be as guys from “Hardspace: Shipbreaker” did — to force the channel configuration in the Wwise.

P.S. Seems like I figured out how to respond with email without extra bloat!

dromer commented:
@eu-ch btw please fix the flake8 test. There seem to be some unused imports that need to be removed :)

eu-ch commented:
Removed the unused import.

Also, I've just added support for 5.1, 7.1, and 7.1.4 :) Thanks for bringing this up, I thought this was useless tbh. So plugins will now gracefully fail if used on a bus with incorrect channel config, instead of outright crashing.

It can still be improved, e.g. if someone wants to use a different configuration, we can consider adding more metadata support, but I think it should be fine as is now, as those are pretty specialised cases.

eu-ch commented:
For history, here's the Wwise project I used for testing
[HvccTest.zip](https://github.com/Wasted-Audio/hvcc/files/13272341/HvccTest.zip)

dromer commented:
Hmm, is there an easy way to build the Windows plugins using mingw?

Otherwise there is no way for me to test the output in the Authoring app using wine :thinking:

eu-ch commented:
I don’t know. Probably possible, but not officially supported, so I wouldn’t waste energy on it. I can bundle test plugins for you to install via launcher and try.

eu-ch commented:
Hey @dromer, I packaged plugins for you to try with the project that was attached earlier.

[HvccTestBug1.zip](https://github.com/Wasted-Audio/hvcc/files/13288596/HvccTestBug1.zip)
[HvccTestFX.zip](https://github.com/Wasted-Audio/hvcc/files/13288597/HvccTestFX.zip)
[HvccTestFX71.zip](https://github.com/Wasted-Audio/hvcc/files/13288598/HvccTestFX71.zip)
[HvccTestSource.zip](https://github.com/Wasted-Audio/hvcc/files/13288599/HvccTestSource.zip)
[HvccTestSource51.zip](https://github.com/Wasted-Audio/hvcc/files/13288600/HvccTestSource51.zip)

You can try installing them via Wwise Launcher like follows:

![image](https://github.com/Wasted-Audio/hvcc/assets/146974283/267469ac-3dae-4525-871c-827511031b18)
![image](https://github.com/Wasted-Audio/hvcc/assets/146974283/ad054d0c-33c7-4145-869d-f10e9e560596)

Please let me know how it goes :)

dromer commented on hvcc/generators/c2wwise/c2wwise.py:
```diff
@@ -43,112 +42,68 @@ def compile(
             if plugin_type == "FX":
-                if num_input_channels > 2:
-                    raise Exception("Wwise FX plugins support a maximum of 2 channels (i.e. [adc~ 1] or [adc~ 1 2]).")
                 if num_input_channels != num_output_channels:
```
Just to confirm: this is still always the case? So it's not possible to have the plugin do any up/down-mixing?
An FX always sits in the path of same number input as output channels.

eu-ch commented:
> So it's not possible to have the plugin do any up/down-mixing?

Correct. Wwise handles this automatically when there’s a mismatch between busses.

eu-ch commented on hvcc/generators/ir2c/static/HvUtils.h:
```diff
@@ -201,7 +201,9 @@
 // Export and Inline
 #if HV_WIN
 #define HV_EXPORT __declspec(dllexport)
+#ifndef __cplusplus // MSVC doesn't like redefining "inline" keyword
```
It won't. This change is only applied to Windows builds. MSVC C/C++ has `inline` keywords already and will refuse to compile if you redefine it.

dromer replied:
Oh right, I missed the `#if HV_WIN` above that :)

eu-ch replied:
I don't think it's a good idea.

First, this is how HV_WIN is defined:
```c
#if _WIN32 || _WIN64 || _MSC_VER
  #define HV_WIN 1
```

So, `#ifndef _MSC_VER` will always be false on Windows. Which leads to the second point: MSVC compiles both C and C++ code. This means your suggestion will disable inline define for C source files on MSVC.

Third, the only reason to define `inline` keyword is to aim at C89 compiler, as this keyword. I can imagine someone building libraries and limiting themselves to this standard for archaic compatibility. Can't say the same about C++.

Fourth, this line defines `inline` to `__inline` which is an MSVC-specific keyword. So, if you're compiling for Windows using MinGW (which is GCC, right?) **and** targeting C89, you may need to redefine `inline` as `__inline__`. That's regarding your MinGW concern. 

But the main point is that there's no reason to define an `inline` macro for C++, and my modification is the smallest and safest change required to solve the MSVC compilation problem.

dromer replied:
oh, right. my bad. I don't nearly build for windows often enough /o\

This was actually from my own mingw fixes haha: https://github.com/Wasted-Audio/hvcc/commit/2c94d5aa204c977ebc5746a342db757ee37783f4

I was actually grepping through the code and OWL still uses an old version of `HvUtils.h` which distracted me.

So, sorry about the noise, carry on.

dromer commented on hvcc/generators/c2wwise/c2wwise.py:
```diff
@@ -31,123 +31,79 @@ class c2wwise:
-        num_output_channels: int = 0,
-        copyright: Optional[str] = None,
-        verbose: Optional[bool] = False
+            cls,
```
maybe silly, but why the additional indentation? :)

eu-ch replied:
Auto formatting :) PyCharm does it according to PEP 8.

dromer replied:
PyCharm does it according to whatever they want. The current code passes all PEP8 checks already :)

I'd be careful with full "auto-formatted" code though, it can also break style guides.

The only auto-formatting I typically do is removal of extraneous white space and addition of ending newlines.
